27 add formatting script for manually reviewed permits

Damonamajor opened the pull request:
This provides two sets of files, and R folder which does the legacy files and then a python script which works for forward moving scripts. I took out the functions which seem to be used in multiple scripts and added them to helpers.py or helpers.R. These can be re-added to the script for the python script, or we can factor out the original permit cleaning script.

Some notes
- If we are doing condos, adding 0000 suffixes may not be the best structure.
- I really don't trust the color matching. You can see that one of them uses a color which is just a theme. I'd be skeptical that these stay identical over time. If someone manages to use the hex color for this it wouldn't be caught.
- I also added another round of checking for errors which would block upload. I didn't fix them so we can verify that they are caught and then probably manually input them. 

xlsx files are uploaded temporarily in order to facilitate PR review, but they can be taken down after.
There are a couple edits I did manually to them, formatting the dates to remove the second suffixes which made them not work, renamed some sheet names (PIN ERRORS-FINAL to PIN ERRORS) and unhiding columns which didn't allow them to be recognized.

I also looked at the output xlsx file for not uploading and there were no non-red PINs.

Damonamajor commented on chicago/legacy_permits/helper.R:
```diff
@@ -0,0 +1,121 @@
+  df[, column_order, drop = FALSE]
+}
+
+read_xlsx_all_char <- function(path, sheet) {
```
This was needed because pivots woul result in different column types.

Damonamajor commented on chicago/format_reviewed_permits_for_upload.py:
```diff
@@ -0,0 +1,227 @@
+import pandas as pd
+
+# Helper function that can be factored out for other scripts
+from helper import (
```
I kept these out since it seems like we may want these in other scripts in this folder.

jeancochrane commented on chicago/legacy_permits/2021/format_legacy_2021.R:
```diff
@@ -0,0 +1,48 @@
+library(openxlsx)
+library(tidyr)
+
+source("legacy_permits/helper.R")
```
**[Nitpick, optional]** Since the R project lives in the `legacy_permits/` subdirectory, I think we should assume that the user's working directory is `legacy_permits/`, meaning we don't need to also list it in paths like this one.

```suggestion
source("helper.R")
```

jeancochrane commented on chicago/legacy_permits/helper.R:
```diff
@@ -0,0 +1,121 @@
+  "Issue Date* [PERMDT]",
+  "Desc 1* [DESC1]",
+  "Desc 2 Code 1 [USER6]",
+  "Desc 2 Code 1 [USER6]2",
```
**[Nitpick, required]** I think this column is a dupe of the previous one, is that right?


```suggestion
```

jeancochrane commented on chicago/legacy_permits/helper.R:
```diff
@@ -0,0 +1,121 @@
+  "Assessable [IS_ASSESS]",
+  "Applicant Street Address* [ADDR1]",
+  "Applicant Address 2 [ADDR2]",
+  "SUFFIX",
```
**[Nitpick, required]** I don't think we want `SUFFIX` in the output, do we?


```suggestion
```

jeancochrane commented on chicago/legacy_permits/helper.R:
```diff
@@ -0,0 +1,121 @@
+      valid_addr_len = nchar(.data[["Applicant Street Address* [ADDR1]"]]) <= 40,
+      valid_note_len = nchar(.data[["Notes [NOTE1]"]]) <= 2000,
+      valid_name_len = nchar(.data[["Applicant* [USER21]"]]) <= 50,
+      valid_amount = .data[["Amount* [AMOUNT]"]] < 2147483647,
```
**[Question, blocking]** I don't totally get what's going on, but something about this check is causing two 2022 Actionable sheet rows to fail this check, even though the amounts look fine and they're less than this number. Any idea what's going on there?

Damonamajor replied:
One value has PIN1 which has an extra 0 in the suffix. It should run through with the other column.

<img width="301" height="21" alt="Image" src="https://github.com/user-attachments/assets/2db3c073-672d-4ca5-a482-843c587ec562" />


And the other is more than 50 characters for the name

<img width="438" height="33" alt="Image" src="https://github.com/user-attachments/assets/282133e8-2554-4737-966f-eec1a80dfecc" />

Damonamajor replied:
Regardless, they are both easy to manually change.

jeancochrane replied:
I no longer see the two rows that were failing the `Amount` check in 2022, so I'm guessing it was related to the `Amount` column being a character before you started casting it to numeric in the `valid_amount` check. All looks good now.

jeancochrane commented on chicago/renv/.gitignore:
```diff
@@ -0,0 +1,7 @@
+lock/
+python/
+sandbox/
+staging/
```
**[Nitpick, required]** I don't think we need this file since `chicago/legacy_permits/` is the main R project directory, right?

jeancochrane commented on chicago/format_reviewed_permits_for_upload.py:
```diff
@@ -0,0 +1,227 @@
+            pin_cell = row_cells[pin_idx]
+
+        if not pin_cell_matches_flag(pin_cell):
+            continue
```
**[Question, non-blocking]** Why are we removing cells with the wrong highlight here? Isn't that the responsibility of `remove_flagged_rows_from_original_xlsx()`?

Damonamajor replied:
The point of this was to have an explicit file to look at with the rows that we were removing for QC. This isn't the best structure, but I wanted to be able to look at the originally formatted xlsx formatting to make sure the color schema was working properly.

jeancochrane commented on chicago/format_reviewed_permits_for_upload.py:
```diff
@@ -0,0 +1,227 @@
+        w.writerow(REQUIRED_COLS)
+        current_lline = 1
+        print(f"Created upload batch: {upload_path}")
+        return w, f, upload_path
```
**[Question, non-blocking]** I don't really get what this function is encapsulating. It seems like we only run it once; I also don't get why we need to write the first header row for the first batch separate from the batch loop below. Do we really need this function?

jeancochrane commented on chicago/format_reviewed_permits_for_upload.py:
```diff
@@ -0,0 +1,227 @@
+            batch_writer = csv.writer(batch_file)
+            batch_writer.writerow(REQUIRED_COLS)
+            for row in batch.itertuples(index=False, name=None):
+                batch_writer.writerow(list(row))
```
**[Suggestion, non-blocking]** Could we use `DataFrame.to_csv()` here to match how we do it for the `need_review_df` above?

jeancochrane commented on chicago/format_reviewed_permits_for_upload.py:
```diff
@@ -0,0 +1,227 @@
+                batch_writer.writerow(list(row))
+
+        print(f"Upload batch saved to: {upload_batch_path}")
+        rows_in_batch += len(batch)
```
**[Question, non-blocking]** What is `rows_in_batch` doing for us? We never seem to use it, aside from initializing it and incrementing it.

jeancochrane commented on chicago/format_reviewed_permits_for_upload.py:
```diff
@@ -0,0 +1,227 @@
+    cleaned_xlsx_path = os.path.join(
+        output_folder, "cleaned_flagged_rows_removed.xlsx"
+    )
+    os.rename(removed_path, cleaned_xlsx_path)
```
**[Question, non-blocking]** Why save this to a different path in `remove_flagged_rows_from_original_xlsx()` and then move it right after saving it? Could we just save it to the final path in the context of the function?

Damonamajor commented on .gitignore:
```diff
@@ -163,3 +163,4 @@ cython_debug/
 chicago/files_for_*
 chicago/chicago_pin_universe*.csv
 
+renv/library/
```
When I'm working in positron and switch base folders to the python script rather than legacy, I get a bunch of things without it.

<img width="344" height="423" alt="Image" src="https://github.com/user-attachments/assets/922be18f-3e35-4874-8725-0bbab50c983b" />

jeancochrane replied:
**[Niptick, optional]** Quick reminder of this thread: You should be able to delete `~/extract-permits/renv/` and then remove this line.


```suggestion
```

Damonamajor commented:
This should fix the key issues regarding the data outputs. We can work on cleaning up the script (especially the python one, once the initial batch is delivered).

Damonamajor commented:
<img width="1240" height="826" alt="image" src="https://github.com/user-attachments/assets/f9abc6df-b302-4cd3-959b-1849eae2679b" />
Not exactly sure about the pip install here. Says there are issues, but it doesn't seem to be having issues locally.

Damonamajor commented on chicago/legacy_permits/2021/format_legacy_2021.R:
```diff
@@ -42,16 +43,22 @@ need_worked <- read_xlsx_all_char(
   mutate(`PIN* [PARID]` = coalesce((meta_pin), (`PIN* [PARID]`))) %>%
   select(-meta_pin) %>%
+   group_by(
+    `PIN* [PARID]`,
```
This deduplicates PINs. The most common cuplrit for this was OHare which had many pins in the `PIN `or `parid `column, but a single pin in the `PIN1 `column. It was maybe 5 observations.

Damonamajor commented:
@jeancochrane I needed to do a bit more clean-up to the file. I found some other oddities when running the new import.

- The previous file they wanted us to upload only had Pin Errors, so I created folders for both Pin Errors and Other Errors.
- There were non-numeric characters which were causing the join to fail with vw_pin_universe so I modified that filter.
- Added a test for de-duplication.
As a note, it won't run when no files are highlighted yellow at the moment. We should fix that, but it doesn't seem urgent, just something to note. 

sometime this was pinged for re-review, but it's good to go again.
